### PR TITLE
[Feature] Discord Webhook Functionality

### PR DESCRIPTION
**Pull request by jonezy35**
## 📃 Description

This PR adds functionality to support Discord webhook notifications. Closes Discord channel in #40 and satisfies user requests in #754 

### ➕ Added

- Discord option in UI for notifications
- Discord listener for speedtest notifications on all runs
- Discord listener for speedtest notifications on thresholds being met

### ✏️ Changed

N/A

### 🔧 Fixed

N/A

### 🗑️ Removed

N/A

## 📷 Screenshots

discord_speedtest

### CODE REVIEW

**Comment by masterwishx** [comment]
@alexjustesen can it be merged now ?

File: app/Filament/Pages/Settings/NotificationPage.php [hunk 1]
```diff
@@ -175,6 +176,48 @@ public function form(Form $form): Form
                                         'default' => 1,
                                         'md' => 2,
                                     ]),
+
```
Comment (alexjustesen) [inline app/Filament/Pages/Settings/NotificationPage.php:179]: Can you move the section up to below "Database", want to keep these in alphabetical order.

File: app/Listeners/SpeedtestCompletedListener.php [hunk 2]
```diff
@@ -75,6 +76,26 @@ public function handle(ResultCreated $event): void
...
+                                     "\nResult ID: " . $event->result->id .
+                                     "\nSite Name: " . $this->generalSettings->site_name .
+                                     "\nPing: " . $event->result->ping . " ms" .
+                                     "\nDownload: " . ($event->result->downloadBits / 1000000) . ' (Mbps)' .
+                                     "\nUpload: " . ($event->result->uploadBits / 1000000) . ' (Mbps)',
+                    ];
+        
+                    // Send the request using Laravel's HTTP client
+                    $response = Http::post($webhook['discord_webhook_url'], $payload);
+
```
Comment (alexjustesen) [inline app/Listeners/SpeedtestCompletedListener.php:94]: Please remove this extra line

File: app/Listeners/Threshold/AbsoluteListener.php [hunk 3]
```diff
@@ -63,6 +63,10 @@ public function handle(ResultCreated $event): void
             $this->telegramChannel($event);
         }
 
+        // Discord notification channel
+        if ($this->notificationSettings->discord_enabled == true && $this->notificationSettings->discord_on_threshold_failure == true) {
+            $this->discordChannel($event);
+        }
```
Comment (alexjustesen) [inline app/Listeners/Threshold/AbsoluteListener.php:69]: Add an extra line after the if statement to give it some breathing room

File: app/Listeners/Threshold/AbsoluteListener.php [hunk 4]
```diff
@@ -219,6 +223,54 @@ protected function telegramChannel(ResultCreated $event): void
...
+                foreach ($this->notificationSettings->discord_webhooks as $webhook) {
+                    // Construct the payload with the failed thresholds information
+                    $contentLines = [
+                        "Result ID: " . $event->result->id,
+                        "Site Name: " . $this->generalSettings->site_name
+                    ];
+                    foreach ($failedThresholds as $metric => $result) {
+                        $contentLines[] = "{$metric} threshold failed with result: {$result}.";
+                    }
+                    $payload = [
```
Comment (alexjustesen) [inline app/Listeners/Threshold/AbsoluteListener.php:262]: Add an extra line before assigning the payload, I'm a fan of code breathing room.

File: app/Listeners/Threshold/AbsoluteListener.php [hunk 5]
```diff
@@ -219,6 +223,54 @@ protected function telegramChannel(ResultCreated $event): void
...
+        // Proceed with sending notifications only if there are any failed thresholds
+        if (count($failedThresholds) > 0) {
+            if ($this->notificationSettings->discord_on_threshold_failure && count($this->notificationSettings->discord_webhooks)) {
+                foreach ($this->notificationSettings->discord_webhooks as $webhook) {
+                    // Construct the payload with the failed thresholds information
+                    $contentLines = [
+                        "Result ID: " . $event->result->id,
+                        "Site Name: " . $this->generalSettings->site_name
+                    ];
+                    foreach ($failedThresholds as $metric => $result) {
```
Comment (alexjustesen) [inline app/Listeners/Threshold/AbsoluteListener.php:259]: Add an extra line before the foreach loop

File: database/migrations/2024_02_19_000000_create_discord_notification_settings.php [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+     * Run the migrations.
+     */
+    public function up(): void
+    {
+        $this->migrator->add('notification.discord_enabled', false);
+        $this->migrator->add('notification.discord_on_speedtest_run', false);
+        $this->migrator->add('notification.discord_on_threshold_failure', false);
+        $this->migrator->add('notification.discord_webhooks', null);
+    }
+};
```
Comment (alexjustesen) [inline database/migrations/2024_02_19_000000_create_discord_notification_settings.php:17]: Need an extra line at the end of the file

**Review by alexjustesen — CHANGES_REQUESTED** [review]

**Comment by alexjustesen** [comment]
@jonezy35 nice work! I left some formatting comments and looks like the Linter wants some additional changes.

If you want to lint locally run `sail bin duster lint --using=pint -v` before committing changes.

**Comment by jonezy35** [comment]
> @jonezy35 nice work! I left some formatting comments and looks like the Linter wants some additional changes.
> 
> If you want to lint locally run `sail bin duster lint --using=pint -v` before committing changes.

@alexjustesen all issues should be fixed now. Made the syntax changes you requested and resolved all lint errors locally before committing. Thanks for the local lint syntax

**Comment by jonezy35** [comment]
Might think about adding the lint check to the development guide so in the future those errors can be handled before the PR is opened

**Comment by alexjustesen** [comment]
Good call, I'll do that and review the PR again tonight.

**Review by alexjustesen — APPROVED** [review]

**Comment by jonezy35** [comment]
@alexjustesen thanks! I'm going to start looking into the slack functionality as well now that I have this down. What is the process/ timeline for the docker container to update with the new functionality once a PR is merged?

**Comment by alexjustesen** [comment]
@jonezy35 going to move this chat to a discussion, got some additional thoughts and changes I need to make and don't want to lose it in the PR.

**Comment by alexjustesen** [comment]
Let's chat here: https://github.com/alexjustesen/speedtest-tracker/discussions/1220

**Pull request merged**